# Building the model

The purpose of this notebook is to build the model that will eventually be used as the model behind the PL prediction for the app, and building the script to rebuild the model when new player data is introduced weekly

## Imports

In [1]:
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import balanced_accuracy_score
from scikeras.wrappers import KerasRegressor
import tensorflow as tf
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV

## Loading the dataset

In [2]:
combined = pd.read_csv("../final_combined_dataframe.csv")

In [3]:
combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1927 entries, 0 to 1926
Data columns (total 89 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Unnamed: 0                          1927 non-null   int64  
 1   match_id                            1927 non-null   object 
 2   goals                               1927 non-null   float64
 3   assists                             1927 non-null   float64
 4   non_penalty_goals                   1927 non-null   float64
 5   penalties_scored                    1927 non-null   float64
 6   penalties_attempted                 1927 non-null   float64
 7   yellow_cards                        1927 non-null   float64
 8   red_cards                           1927 non-null   float64
 9   expected_goals                      1927 non-null   float64
 10  non_penalty_expected_goals          1927 non-null   float64
 11  expected_assisted_goals             1927 no

In [4]:
combined.head()

Unnamed: 0 match_id     goals   assists  non_penalty_goals  \
0           0  m-00381  0.907776  0.396142           0.738889   
1           1  m-00382  0.155008  0.133838           0.123306   
2           2  m-00383 -0.039926 -0.056540          -0.137227   
3           3  m-00384  0.133878 -0.008924           0.122145   
4           4  m-00385  0.401345  0.138483           0.411797   

   penalties_scored  penalties_attempted  yellow_cards  red_cards  \
0          0.168887             0.236302      0.093022  -0.048172   
1          0.031702             0.039064      0.131687  -0.000329   
2          0.097301             0.109960      0.066208   0.011733   
3          0.011733             0.011733      0.185723  -0.023049   
4         -0.010452            -0.001508      0.071065  -0.018950   

   expected_goals  ...  home_shots_on_target  away_shots_on_target  \
0        0.776569  ...                     6                     4   
1        0.200098  ...                     4                     1   
2       -0.109844  ...                     6                     9   
3        0.212802  ...                     1                     4   
4        0.419927  ...                     2                     5   

   home_corners  away_corners  home_fouls  away_fouls  home_yellow_cards  \
0             2             5          11           8                  2   
1             7             4          11           9                  1   
2             5             5           9          11                  1   
3             2             5           9           8                  2   
4             3             5          11          12                  2   

   away_yellow_cards  home_red_cards  away_red_cards  
0                  1               0               0  
1                  1               0               0  
2                  2               0               0  
3                  1               0               0  
4                  2               0               0  

[5 rows x 89 columns]

## Column Lists

In [5]:
output_columns = [
	"home_goals", "away_goals", "home_shots", "away_shots", "home_shots_on_target", "away_shots_on_target",
	"home_corners", "away_corners", "home_fouls", "away_fouls", "home_yellow_cards", "away_yellow_cards",
	"home_red_cards", "away_red_cards"
]

player_stats_columns = [
	"player_id", "minutes_played","ninetys","goals","assists","non_penalty_goals","penalties_scored","penalties_attempted","yellow_cards","red_cards","expected_goals",
	"non_penalty_expected_goals","expected_assisted_goals","progressive_carries","progressive_passes","progressive_passes_received","total_passing_distance","total_progressive_passing_distance","short_passes_completed","short_passes_attempted","medium_passes_completed","medium_passes_attempted",
	"long_passes_completed","long_passes_attempted","expected_assists","key_passes","passes_into_final_third","passes_into_penalty_area","crosses_into_penalty_area","shots","shots_on_target","average_shot_distance","shots_from_free_kicks",
	"shots_from_penalties","touches","touches_in_defensive_penalty_area","touches_in_defensive_third","touches_in_middle_third","touches_in_attacking_third","touches_in_attacking_penalty_area","live_ball_touches","take_ons_attempted","take_ons_succeeded","times_tackled_during_take_on",
	"carries","total_carrying_distance","progressive_carrying_distance","carries_into_final_third","carries_into_penalty_area","miscontrols","dispossessed","passes_received","tackles","tackles_won","defensive_third_tackles",
	"middle_third_tackles","attacking_third_tackles","dribblers_tackled","dribbler_tackles_attempted","shots_blocked","passes_blocked","interceptions","clearances","errors_leading_to_shot","goals_against","shots_on_target_against","saves","clean_sheets","penalties_faced","penalties_allowed","penalties_saved","penalties_missed"
]
pure_stats_columns = [
	"minutes_played","goals","assists","non_penalty_goals","penalties_scored","penalties_attempted","yellow_cards","red_cards","expected_goals",
	"non_penalty_expected_goals","expected_assisted_goals","progressive_carries","progressive_passes","progressive_passes_received","total_passing_distance","total_progressive_passing_distance","short_passes_completed","short_passes_attempted","medium_passes_completed","medium_passes_attempted",
	"long_passes_completed","long_passes_attempted","expected_assists","key_passes","passes_into_final_third","passes_into_penalty_area","crosses_into_penalty_area","shots","shots_on_target","average_shot_distance","shots_from_free_kicks",
	"shots_from_penalties","touches","touches_in_defensive_penalty_area","touches_in_defensive_third","touches_in_middle_third","touches_in_attacking_third","touches_in_attacking_penalty_area","live_ball_touches","take_ons_attempted","take_ons_succeeded","times_tackled_during_take_on",
	"carries","total_carrying_distance","progressive_carrying_distance","carries_into_final_third","carries_into_penalty_area","miscontrols","dispossessed","passes_received","tackles","tackles_won","defensive_third_tackles",
	"middle_third_tackles","attacking_third_tackles","dribblers_tackled","dribbler_tackles_attempted","shots_blocked","passes_blocked","interceptions","clearances","errors_leading_to_shot","goals_against","shots_on_target_against","saves","clean_sheets","penalties_faced","penalties_allowed","penalties_saved","penalties_missed"
]
team_stats_columns = [
	"team_id", "goals","assists","non_penalty_goals","penalties_scored","penalties_attempted","yellow_cards","red_cards","expected_goals",
	"non_penalty_expected_goals","expected_assisted_goals","progressive_carries","progressive_passes","progressive_passes_received","total_passing_distance","total_progressive_passing_distance","short_passes_completed","short_passes_attempted","medium_passes_completed","medium_passes_attempted",
	"long_passes_completed","long_passes_attempted","expected_assists","key_passes","passes_into_final_third","passes_into_penalty_area","crosses_into_penalty_area","shots","shots_on_target","average_shot_distance","shots_from_free_kicks",
	"shots_from_penalties","touches","touches_in_defensive_penalty_area","touches_in_defensive_third","touches_in_middle_third","touches_in_attacking_third","touches_in_attacking_penalty_area","live_ball_touches","take_ons_attempted","take_ons_succeeded","times_tackled_during_take_on",
	"carries","total_carrying_distance","progressive_carrying_distance","carries_into_final_third","carries_into_penalty_area","miscontrols","dispossessed","passes_received","tackles","tackles_won","defensive_third_tackles",
	"middle_third_tackles","attacking_third_tackles","dribblers_tackled","dribbler_tackles_attempted","shots_blocked","passes_blocked","interceptions","clearances","errors_leading_to_shot","goals_against","shots_on_target_against","saves","clean_sheets","penalties_faced","penalties_allowed","penalties_saved","penalties_missed"
]
pure_stats_columns_no_minutes = [
	"goals","assists","non_penalty_goals","penalties_scored","penalties_attempted","yellow_cards","red_cards","expected_goals",
	"non_penalty_expected_goals","expected_assisted_goals","progressive_carries","progressive_passes","progressive_passes_received","total_passing_distance","total_progressive_passing_distance","short_passes_completed","short_passes_attempted","medium_passes_completed","medium_passes_attempted",
	"long_passes_completed","long_passes_attempted","expected_assists","key_passes","passes_into_final_third","passes_into_penalty_area","crosses_into_penalty_area","shots","shots_on_target","average_shot_distance","shots_from_free_kicks",
	"shots_from_penalties","touches","touches_in_defensive_penalty_area","touches_in_defensive_third","touches_in_middle_third","touches_in_attacking_third","touches_in_attacking_penalty_area","live_ball_touches","take_ons_attempted","take_ons_succeeded","times_tackled_during_take_on",
	"carries","total_carrying_distance","progressive_carrying_distance","carries_into_final_third","carries_into_penalty_area","miscontrols","dispossessed","passes_received","tackles","tackles_won","defensive_third_tackles",
	"middle_third_tackles","attacking_third_tackles","dribblers_tackled","dribbler_tackles_attempted","shots_blocked","passes_blocked","interceptions","clearances","errors_leading_to_shot","goals_against","shots_on_target_against","saves","clean_sheets","penalties_faced","penalties_allowed","penalties_saved","penalties_missed"
]

## Scaling

In [6]:
scaler = StandardScaler(copy=True)

In [7]:
combined_standardized = scaler.fit_transform(combined[pure_stats_columns_no_minutes])

In [8]:
evr = {}
n_components = {}
feature_to_pc = {}

for n in range(1,26):
	pca = PCA(n_components = n, random_state=938)
	pca.fit(combined_standardized)
	feature_to_pc_map = pd.DataFrame(pca.components_, columns=pure_stats_columns_no_minutes)
	components = pca.transform(combined_standardized)
	components_df = pd.DataFrame(data=components[:, [p for p in range(n)]], columns=pca.get_feature_names_out(), )
	
	evr[n] = sum(pca.explained_variance_ratio_)
	n_components[n] = components_df
	feature_to_pc[n] = feature_to_pc_map

pd.DataFrame(data=evr, index=["explained_variance_ratio"]).T

explained_variance_ratio
1                   0.686249
2                   0.801371
3                   0.860920
4                   0.885021
5                   0.902991
6                   0.916988
7                   0.927908
8                   0.938424
9                   0.947595
10                  0.954277
11                  0.960357
12                  0.965957
13                  0.970505
14                  0.974268
15                  0.977187
16                  0.979719
17                  0.982173
18                  0.984039
19                  0.985737
20                  0.987423
21                  0.988736
22                  0.990020
23                  0.991050
24                  0.991912
25                  0.992695

Check the effect of different numbers if PCs on the outcome of the NN, testing n=2, 5 and 10. Use each n components to train and test the neural network, and compare the performance of each one

## Number of Features

In [9]:
INPUT_SIZE = 69
OUTPUT_SIZE = 14
ROWS = 1927

### n = 2

Train test split of data

In [10]:
n=2
X = n_components[n][['pca'+ str(x) for x in range(n)]]
Y = combined[output_columns]

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=629) 

Model definition

In [11]:
model = tf.keras.models.Sequential([
	tf.keras.layers.Dense(32, activation='relu', input_dim=len(X_train.columns)),
	tf.keras.layers.Dense(128, activation='relu'),
	tf.keras.layers.Dense(OUTPUT_SIZE)
])
model.compile(loss="mean_squared_error", optimizer="sgd", metrics="accuracy")
model.fit(X_train, y_train, epochs=500, batch_size=50);

Epoch 1/500
31/31 [==============================] - 0s 483us/step - loss: 28.4828 - accuracy: 0.3796
Epoch 2/500
31/31 [==============================] - 0s 317us/step - loss: 16.9982 - accuracy: 0.4607
Epoch 3/500
31/31 [==============================] - 0s 376us/step - loss: 13.8711 - accuracy: 0.4711
Epoch 4/500
 1/31 [..............................] - ETA: 0s - loss: 13.6099 - accuracy: 0.5600

2024-01-02 10:19:28.333353: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


31/31 [==============================] - 0s 341us/step - loss: 11.1312 - accuracy: 0.4698
Epoch 5/500
31/31 [==============================] - 0s 342us/step - loss: 9.4545 - accuracy: 0.4705
Epoch 6/500
31/31 [==============================] - 0s 393us/step - loss: 9.2953 - accuracy: 0.4698
Epoch 7/500
31/31 [==============================] - 0s 335us/step - loss: 8.6533 - accuracy: 0.4757
Epoch 8/500
31/31 [==============================] - 0s 320us/step - loss: 9.0268 - accuracy: 0.4763
Epoch 9/500
31/31 [==============================] - 0s 362us/step - loss: 8.6638 - accuracy: 0.4763
Epoch 10/500
31/31 [==============================] - 0s 348us/step - loss: 8.5361 - accuracy: 0.4711
Epoch 11/500
31/31 [==============================] - 0s 398us/step - loss: 8.4729 - accuracy: 0.4744
Epoch 12/500
31/31 [==============================] - 0s 336us/step - loss: 8.5897 - accuracy: 0.4783
Epoch 13/500
31/31 [==============================] - 0s 329us/step - loss: 8.4086 - accuracy: 0.47

In [12]:
y_hat = model.predict(X_test)
print(y_hat[1])
y_test.head()

13/13 [==============================] - 0s 283us/step
[ 1.5740263   1.1674742  12.788379   11.963009    4.783677    4.2019944
  5.076368    4.7898192  10.231185   10.893661    1.4929272   1.8858938
  0.17641644  0.08949739]


home_goals  away_goals  home_shots  away_shots  home_shots_on_target  \
1489           1           5          13          13                     5   
643            0           0          21           8                     4   
1263           1           2           9          17                     3   
848            1           3          19          15                     6   
95             0           4           7          24                     1   

      away_shots_on_target  home_corners  away_corners  home_fouls  \
1489                     7             7             3           4   
643                      1            11             7          11   
1263                     7             4            11           8   
848                      6             5             4           9   
95                       8             4             4          14   

      away_fouls  home_yellow_cards  away_yellow_cards  home_red_cards  \
1489           6                  0                  1               0   
643            9                  4                  3               0   
1263          12                  0                  0               0   
848           10                  0                  0               0   
95            10                  4                  2               0   

      away_red_cards  
1489               0  
643                0  
1263               0  
848                0  
95                 0

### n = 5

In [13]:
n=5
X = n_components[n][['pca'+ str(x) for x in range(n)]]
Y = combined[output_columns]

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=629) 

In [14]:
model = tf.keras.models.Sequential([
	tf.keras.layers.Dense(32, activation='relu', input_dim=len(X_train.columns)),
	tf.keras.layers.Dense(128, activation='relu'),
	tf.keras.layers.Dense(OUTPUT_SIZE)
])
model.compile(loss="mean_squared_error", optimizer="sgd", metrics="accuracy")
model.fit(X_train, y_train, epochs=500, batch_size=50);

Epoch 1/500
31/31 [==============================] - 0s 506us/step - loss: 28.6890 - accuracy: 0.4127
Epoch 2/500
31/31 [==============================] - 0s 346us/step - loss: 14.9569 - accuracy: 0.4711
Epoch 3/500
31/31 [==============================] - 0s 321us/step - loss: 12.1368 - accuracy: 0.4705
Epoch 4/500
31/31 [==============================] - 0s 329us/step - loss: 10.6339 - accuracy: 0.4705
Epoch 5/500
31/31 [==============================] - 0s 326us/step - loss: 9.5916 - accuracy: 0.4724
Epoch 6/500
31/31 [==============================] - 0s 324us/step - loss: 9.5993 - accuracy: 0.4731
Epoch 7/500
31/31 [==============================] - 0s 384us/step - loss: 8.9417 - accuracy: 0.4737
Epoch 8/500
31/31 [==============================] - 0s 291us/step - loss: 8.8603 - accuracy: 0.4711
Epoch 9/500
31/31 [==============================] - 0s 300us/step - loss: 8.6751 - accuracy: 0.4711
Epoch 10/500
31/31 [==============================] - 0s 359us/step - loss: 8.7066 - ac

In [15]:
y_hat = model.predict(X_test)
print(y_hat[1])
y_test.head()

13/13 [==============================] - 0s 320us/step
[ 1.6304392   1.3572861  14.205526   11.4678      5.1762853   3.9137225
  4.8312798   4.643898    9.541681   11.740981    1.0904438   1.8678386
  0.19181609  0.06824517]


home_goals  away_goals  home_shots  away_shots  home_shots_on_target  \
1489           1           5          13          13                     5   
643            0           0          21           8                     4   
1263           1           2           9          17                     3   
848            1           3          19          15                     6   
95             0           4           7          24                     1   

      away_shots_on_target  home_corners  away_corners  home_fouls  \
1489                     7             7             3           4   
643                      1            11             7          11   
1263                     7             4            11           8   
848                      6             5             4           9   
95                       8             4             4          14   

      away_fouls  home_yellow_cards  away_yellow_cards  home_red_cards  \
1489           6                  0                  1               0   
643            9                  4                  3               0   
1263          12                  0                  0               0   
848           10                  0                  0               0   
95            10                  4                  2               0   

      away_red_cards  
1489               0  
643                0  
1263               0  
848                0  
95                 0

### n = 10

In [16]:
n=10
X = n_components[n][['pca'+ str(x) for x in range(n)]]
Y = combined[output_columns]

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=629) 

In [17]:
model = tf.keras.models.Sequential([
	tf.keras.layers.Dense(32, activation='relu', input_dim=len(X_train.columns)),
	tf.keras.layers.Dense(128, activation='relu'),
	tf.keras.layers.Dense(OUTPUT_SIZE)
])
model.compile(loss="mean_squared_error", optimizer="sgd", metrics="accuracy")
model.fit(X_train, y_train, epochs=500, batch_size=50);

Epoch 1/500
31/31 [==============================] - 0s 436us/step - loss: 26.1294 - accuracy: 0.3686
Epoch 2/500
31/31 [==============================] - 0s 379us/step - loss: 14.5856 - accuracy: 0.4698
Epoch 3/500
31/31 [==============================] - 0s 390us/step - loss: 12.1116 - accuracy: 0.4711
Epoch 4/500
31/31 [==============================] - 0s 311us/step - loss: 10.6926 - accuracy: 0.4718
Epoch 5/500
31/31 [==============================] - 0s 365us/step - loss: 9.8622 - accuracy: 0.4685
Epoch 6/500
31/31 [==============================] - 0s 335us/step - loss: 9.3550 - accuracy: 0.4685
Epoch 7/500
31/31 [==============================] - 0s 313us/step - loss: 9.0886 - accuracy: 0.4679
Epoch 8/500
31/31 [==============================] - 0s 310us/step - loss: 8.8962 - accuracy: 0.4724
Epoch 9/500
31/31 [==============================] - 0s 311us/step - loss: 8.7069 - accuracy: 0.4705
Epoch 10/500
31/31 [==============================] - 0s 322us/step - loss: 8.6472 - ac

In [18]:
y_hat = model.predict(X_test)
print(y_hat[1])
y_test.head()

13/13 [==============================] - 0s 310us/step
[ 2.2252014   0.90585923 17.664309    8.731304    6.0318093   3.362132
  7.5312057   2.8790896  10.017103   12.953475    1.3428208   2.1706443
  0.11901493  0.06079175]


home_goals  away_goals  home_shots  away_shots  home_shots_on_target  \
1489           1           5          13          13                     5   
643            0           0          21           8                     4   
1263           1           2           9          17                     3   
848            1           3          19          15                     6   
95             0           4           7          24                     1   

      away_shots_on_target  home_corners  away_corners  home_fouls  \
1489                     7             7             3           4   
643                      1            11             7          11   
1263                     7             4            11           8   
848                      6             5             4           9   
95                       8             4             4          14   

      away_fouls  home_yellow_cards  away_yellow_cards  home_red_cards  \
1489           6                  0                  1               0   
643            9                  4                  3               0   
1263          12                  0                  0               0   
848           10                  0                  0               0   
95            10                  4                  2               0   

      away_red_cards  
1489               0  
643                0  
1263               0  
848                0  
95                 0

### n=15

In [19]:
n=15
X = n_components[n][['pca'+ str(x) for x in range(n)]]
Y = combined[output_columns]

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=629) 

In [20]:
model = tf.keras.models.Sequential([
	tf.keras.layers.Dense(32, activation='relu', input_dim=len(X_train.columns)),
	tf.keras.layers.Dense(128, activation='relu'),
	tf.keras.layers.Dense(OUTPUT_SIZE)
])
model.compile(loss="mean_squared_error", optimizer="sgd", metrics="accuracy")
model.fit(X_train, y_train, epochs=500, batch_size=50);

Epoch 1/500
31/31 [==============================] - 0s 402us/step - loss: 27.0578 - accuracy: 0.3842
Epoch 2/500
31/31 [==============================] - 0s 383us/step - loss: 14.3672 - accuracy: 0.4685
Epoch 3/500
31/31 [==============================] - 0s 315us/step - loss: 11.9893 - accuracy: 0.4698
Epoch 4/500
31/31 [==============================] - 0s 305us/step - loss: 10.7316 - accuracy: 0.4698
Epoch 5/500
31/31 [==============================] - 0s 326us/step - loss: 9.9383 - accuracy: 0.4679
Epoch 6/500
31/31 [==============================] - 0s 345us/step - loss: 9.4515 - accuracy: 0.4685
Epoch 7/500
31/31 [==============================] - 0s 362us/step - loss: 9.1484 - accuracy: 0.4705
Epoch 8/500
31/31 [==============================] - 0s 322us/step - loss: 9.0275 - accuracy: 0.4698
Epoch 9/500
31/31 [==============================] - 0s 319us/step - loss: 8.7714 - accuracy: 0.4698
Epoch 10/500
31/31 [==============================] - 0s 353us/step - loss: 8.7262 - ac

In [21]:
y_hat = model.predict(X_test)
print(y_hat[1])
y_test.head()

13/13 [==============================] - 0s 404us/step
[1.6257322e+00 1.0657034e+00 1.5229052e+01 7.9977889e+00 4.1338181e+00
 2.1900196e+00 6.1345773e+00 3.6977003e+00 9.9363012e+00 1.1320057e+01
 1.3569801e+00 1.6203835e+00 2.5361693e-03 9.0031125e-02]


home_goals  away_goals  home_shots  away_shots  home_shots_on_target  \
1489           1           5          13          13                     5   
643            0           0          21           8                     4   
1263           1           2           9          17                     3   
848            1           3          19          15                     6   
95             0           4           7          24                     1   

      away_shots_on_target  home_corners  away_corners  home_fouls  \
1489                     7             7             3           4   
643                      1            11             7          11   
1263                     7             4            11           8   
848                      6             5             4           9   
95                       8             4             4          14   

      away_fouls  home_yellow_cards  away_yellow_cards  home_red_cards  \
1489           6                  0                  1               0   
643            9                  4                  3               0   
1263          12                  0                  0               0   
848           10                  0                  0               0   
95            10                  4                  2               0   

      away_red_cards  
1489               0  
643                0  
1263               0  
848                0  
95                 0

### n=20

In [22]:
n=20
X = n_components[n][['pca'+ str(x) for x in range(n)]]
Y = combined[output_columns]

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=629) 

In [23]:
model = tf.keras.models.Sequential([
	tf.keras.layers.Dense(32, activation='relu', input_dim=len(X_train.columns)),
	tf.keras.layers.Dense(128, activation='relu'),
	tf.keras.layers.Dense(OUTPUT_SIZE)
])
model.compile(loss="mean_squared_error", optimizer="sgd", metrics="accuracy")
model.fit(X_train, y_train, epochs=500, batch_size=50);

Epoch 1/500
31/31 [==============================] - 0s 385us/step - loss: 27.9933 - accuracy: 0.4244
Epoch 2/500
31/31 [==============================] - 0s 382us/step - loss: 13.9544 - accuracy: 0.4640
Epoch 3/500
31/31 [==============================] - 0s 427us/step - loss: 11.6452 - accuracy: 0.4724
Epoch 4/500
31/31 [==============================] - 0s 328us/step - loss: 10.5254 - accuracy: 0.4672
Epoch 5/500
31/31 [==============================] - 0s 387us/step - loss: 9.7614 - accuracy: 0.4679
Epoch 6/500
31/31 [==============================] - 0s 323us/step - loss: 9.3488 - accuracy: 0.4692
Epoch 7/500
31/31 [==============================] - 0s 350us/step - loss: 9.1894 - accuracy: 0.4705
Epoch 8/500
31/31 [==============================] - 0s 351us/step - loss: 8.8515 - accuracy: 0.4692
Epoch 9/500
31/31 [==============================] - 0s 315us/step - loss: 8.8603 - accuracy: 0.4692
Epoch 10/500
31/31 [==============================] - 0s 320us/step - loss: 8.7357 - ac

In [24]:
y_hat = model.predict(X_test)
print(y_hat[1])
y_test.head()

13/13 [==============================] - 0s 346us/step
[ 1.0493637   1.6811217  10.038267   13.1147      3.5899231   4.6082892
  4.13195     4.226933   11.587735   11.591479    1.8244305   1.4548976
  0.26080847  0.19176434]


home_goals  away_goals  home_shots  away_shots  home_shots_on_target  \
1489           1           5          13          13                     5   
643            0           0          21           8                     4   
1263           1           2           9          17                     3   
848            1           3          19          15                     6   
95             0           4           7          24                     1   

      away_shots_on_target  home_corners  away_corners  home_fouls  \
1489                     7             7             3           4   
643                      1            11             7          11   
1263                     7             4            11           8   
848                      6             5             4           9   
95                       8             4             4          14   

      away_fouls  home_yellow_cards  away_yellow_cards  home_red_cards  \
1489           6                  0                  1               0   
643            9                  4                  3               0   
1263          12                  0                  0               0   
848           10                  0                  0               0   
95            10                  4                  2               0   

      away_red_cards  
1489               0  
643                0  
1263               0  
848                0  
95                 0

The above experiments show us the accuracy of each model increases with an increased number of features up to a point, it seem somewhere between 15 and 20 features, where the accuracy on the training data peaks. I will use 15 principal components going forward for the investigations, as this seems to yield the highest accuracy of those tested here

In terms of predicting exact values for the match facts, I am not expecting exact matches between model and the actual results, just hoping for a close approximation, or at the very least, the correct side with the higher value, i.e. home team scores more goals than the away team is correct

## 15 PC model build

#### Splitting data

In [25]:
X = combined[pure_stats_columns_no_minutes]
y = combined[output_columns]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=543)

#### Scaling X and y

In [26]:
X_scaler = StandardScaler(copy=True).fit(X_train)
y_scaler = StandardScaler(copy=True).fit(y_train)

X_train = X_scaler.transform(X_train)
# y_train = y_scaler.transform(y_train)

X_test = X_scaler.transform(X_test)
# y_test = y_scaler.transform(y_test)

#### Applying PCA

In [27]:
n=15

pca = PCA(n_components = n, random_state=576)
pca.fit(X_train)

X_train = pca.transform(X_train)
X_test = pca.transform(X_test)

The below table shows the breakdown of each principle component in terms of the proportion of the value from each of the table columns used to make it up

In [28]:
df = pd.DataFrame(pca.components_, columns=pure_stats_columns_no_minutes)
df.index.name = "principle_component_number"
df

goals   assists  non_penalty_goals  \
principle_component_number                                          
0                          -0.114827 -0.129052          -0.116702   
1                          -0.076207 -0.074874          -0.096449   
2                           0.129684  0.094786           0.202589   
3                           0.012339 -0.056923           0.047812   
4                           0.393711 -0.057483           0.078656   
5                           0.065291  0.118182          -0.177586   
6                           0.027925  0.089704          -0.001078   
7                          -0.060446  0.059229           0.035370   
8                          -0.015947 -0.021876          -0.129169   
9                           0.153125  0.170773           0.115961   
10                          0.080519  0.033866          -0.094403   
11                          0.085726  0.118866           0.188995   
12                         -0.126072  0.215943           0.027401   
13                         -0.044192 -0.050452          -0.087373   
14                          0.025403 -0.121585           0.026615   

                            penalties_scored  penalties_attempted  \
principle_component_number                                          
0                                  -0.067303            -0.068918   
1                                  -0.099214            -0.100669   
2                                   0.201222             0.203529   
3                                   0.517905             0.509710   
4                                   0.006452            -0.016994   
5                                   0.105684             0.110093   
6                                  -0.024269            -0.035036   
7                                  -0.041391            -0.056498   
8                                   0.147798             0.148286   
9                                  -0.256473            -0.218797   
10                                 -0.066472            -0.073463   
11                                 -0.211855            -0.203311   
12                                  0.122083             0.173108   
13                                  0.050368             0.036710   
14                                 -0.001054             0.006093   

                            yellow_cards  red_cards  expected_goals  \
principle_component_number                                            
0                              -0.133266  -0.078929       -0.119126   
1                               0.002926   0.017051       -0.102086   
2                              -0.091492  -0.078626        0.197089   
3                               0.044948  -0.002697        0.115383   
4                               0.072440  -0.070190        0.100952   
5                              -0.096678  -0.108414       -0.176219   
6                              -0.091844   0.046270       -0.024446   
7                              -0.041428  -0.863648        0.018545   
8                               0.112371  -0.392915       -0.066364   
9                              -0.005597  -0.063845        0.062654   
10                              0.026856   0.147354       -0.064391   
11                              0.052591  -0.029166        0.108949   
12                             -0.100336   0.185364        0.052679   
13                              0.076701   0.000418       -0.046010   
14                              0.017823   0.009800        0.013383   

                            non_penalty_expected_goals  \
principle_component_number                               
0                                            -0.120603   
1                                            -0.096433   
2                                             0.184715   
3                                             0.042584   
4                                             0.114822   
5                                            -0.214140   
6          

#### Model Definition

In [29]:
INPUT_SIZE = len(X_train[0])
OUTPUT_SIZE = len(output_columns)

Define

In [30]:
model = tf.keras.models.Sequential([
	tf.keras.layers.Dense(15, activation='relu'),
	tf.keras.layers.Dense(12, activation='relu'),
	tf.keras.layers.Dense(OUTPUT_SIZE, activation='relu')
])

Compile

In [31]:
model.compile(loss="mean_squared_error", optimizer="sgd", metrics="accuracy")

Fit

In [32]:
model.fit(X_train, y_train, epochs=500, batch_size=25);

Epoch 1/500


62/62 [==============================] - 0s 344us/step - loss: 34.6842 - accuracy: 0.2758
Epoch 2/500
62/62 [==============================] - 0s 318us/step - loss: 13.6334 - accuracy: 0.4705
Epoch 3/500
62/62 [==============================] - 0s 291us/step - loss: 11.2494 - accuracy: 0.4783
Epoch 4/500
62/62 [==============================] - 0s 273us/step - loss: 10.4279 - accuracy: 0.4757
Epoch 5/500
62/62 [==============================] - 0s 273us/step - loss: 10.3266 - accuracy: 0.4770
Epoch 6/500
62/62 [==============================] - 0s 292us/step - loss: 10.0365 - accuracy: 0.4750
Epoch 7/500
62/62 [==============================] - 0s 296us/step - loss: 9.8404 - accuracy: 0.4757
Epoch 8/500
62/62 [==============================] - 0s 269us/step - loss: 9.8838 - accuracy: 0.4757
Epoch 9/500
62/62 [==============================] - 0s 276us/step - loss: 9.7951 - accuracy: 0.4789
Epoch 10/500
62/62 [==============================] - 0s 266us/step - loss: 9.7332 - accuracy: 0.

#### Testing

In [33]:
y_hat = model.predict(X_test)

13/13 [==============================] - 0s 327us/step


In [34]:
total_shots = 0
total_shots_over = 0
total_shots_under = 0

total_shots_on_target = 0
total_shots_on_target_over = 0
total_shots_on_target_under = 0

total_booking_points = 0
total_booking_points_over = 0
total_booking_points_under = 0

correct_score = 0

winner = 0

total_corners = 0
total_corners_over = 0
total_corners_under = 0

total_fouls = 0
total_fouls_over = 0
total_fouls_under = 0

goals_over = 0
goals_under = 0

all_under = 0
all_over = 0

total_tested  = len(y_hat)

In [35]:
for idx, y in enumerate(y_hat):
	home_goals_hat, away_goals_hat, home_shots_hat, away_shots_hat, home_shots_on_target_hat, away_shots_on_target_hat, home_corners_hat, away_corners_hat, home_fouls_hat, away_fouls_hat, home_yellow_cards_hat, away_yellow_cards_hat, home_red_cards_hat, away_red_cards_hat = y
	home_goals, away_goals, home_shots, away_shots, home_shots_on_target, away_shots_on_target, home_corners, away_corners, home_fouls, away_fouls, home_yellow_cards, away_yellow_cards, home_red_cards, away_red_cards = y_test.iloc[idx].tolist()
	
	total_shots += 1 if np.floor(home_shots_hat)+np.floor(away_shots_hat) == home_shots+away_shots else 0
	total_shots_over += 1 if np.floor(home_shots_hat)+np.floor(away_shots_hat) > home_shots+away_shots else 0
	total_shots_under += 1 if np.floor(home_shots_hat)+np.floor(away_shots_hat) < home_shots+away_shots else 0
	
	total_shots_on_target += 1 if np.floor(home_shots_on_target_hat)+np.floor(away_shots_on_target_hat) == home_shots_on_target+away_shots_on_target else 0
	total_shots_on_target_over += 1 if np.floor(home_shots_on_target_hat)+np.floor(away_shots_on_target_hat) > home_shots_on_target+away_shots_on_target else 0
	total_shots_on_target_under += 1 if np.floor(home_shots_on_target_hat)+np.floor(away_shots_on_target_hat) < home_shots_on_target+away_shots_on_target else 0

	total_booking_points += 1 if (np.floor(home_yellow_cards_hat)+np.floor(away_yellow_cards_hat))*10+(np.floor(home_red_cards_hat)+np.floor(away_red_cards_hat))*25 == (home_yellow_cards+away_yellow_cards)*10+(home_red_cards+away_red_cards)*25 else 0
	total_booking_points_over += 1 if (np.floor(home_yellow_cards_hat)+np.floor(away_yellow_cards_hat))*10+(np.floor(home_red_cards_hat)+np.floor(away_red_cards_hat))*25 > (home_yellow_cards+away_yellow_cards)*10+(home_red_cards+away_red_cards)*25 else 0
	total_booking_points_under += 1 if (np.floor(home_yellow_cards_hat)+np.floor(away_yellow_cards_hat))*10+(np.floor(home_red_cards_hat)+np.floor(away_red_cards_hat))*25 < (home_yellow_cards+away_yellow_cards)*10+(home_red_cards+away_red_cards)*25 else 0

	correct_score += 1 if np.floor(home_goals_hat) == home_goals and np.floor(away_goals_hat) == away_goals else 0
	
	winner += 1 if (home_goals_hat > away_goals_hat and home_goals > away_goals) or (home_goals_hat < away_goals_hat and home_goals < away_goals) else 0

	total_fouls += 1 if np.floor(home_fouls_hat) + np.floor(away_fouls_hat) == home_fouls + away_fouls else 0
	total_fouls_over += 1 if np.floor(home_fouls_hat) + np.floor(away_fouls_hat) > home_fouls+ away_fouls else 0
	total_fouls_under += 1 if np.floor(home_fouls_hat) + np.floor(away_fouls_hat) < home_fouls+ away_fouls else 0

	total_corners += 1 if np.floor(home_corners_hat) + np.floor(away_corners_hat) == home_corners + away_corners else 0
	total_corners_over += 1 if np.floor(home_corners_hat) + np.floor(away_corners_hat) > home_corners + away_corners else 0
	total_corners_under += 1 if np.floor(home_corners_hat) + np.floor(away_corners_hat) < home_corners + away_corners else 0

	goals_over += 1 if np.floor(home_goals_hat) + np.floor(away_goals_hat) > home_goals + away_goals else 0
	goals_under += 1 if np.floor(home_goals_hat) + np.floor(away_goals_hat) < home_goals + away_goals else 0

	all_under += 1 if np.floor(home_shots_hat)+np.floor(away_shots_hat) < home_shots+away_shots and np.floor(home_shots_on_target_hat)+np.floor(away_shots_on_target_hat) < home_shots_on_target+away_shots_on_target and (np.floor(home_yellow_cards_hat)+np.floor(away_yellow_cards_hat))*10+(np.floor(home_red_cards_hat)+np.floor(away_red_cards_hat))*25 < (home_yellow_cards+away_yellow_cards)*10+(home_red_cards+away_red_cards)*25 and np.floor(home_corners_hat) + np.floor(away_corners_hat) < home_corners + away_corners and np.floor(home_corners_hat) + np.floor(away_corners_hat) < home_corners + away_corners and np.floor(home_goals_hat) + np.floor(away_goals_hat) < home_goals + away_goals else 0
	all_over += 1 if np.floor(home_shots_hat)+np.floor(away_shots_hat) > home_shots+away_shots and np.floor(home_shots_on_target_hat)+np.floor(away_shots_on_target_hat) > home_shots_on_target+away_shots_on_target and (np.floor(home_yellow_cards_hat)+np.floor(away_yellow_cards_hat))*10+(np.floor(home_red_cards_hat)+np.floor(away_red_cards_hat))*25 > (home_yellow_cards+away_yellow_cards)*10+(home_red_cards+away_red_cards)*25 and np.floor(home_corners_hat) + np.floor(away_corners_hat) > home_corners + away_corners and np.floor(home_goals_hat) + np.floor(away_goals_hat) > home_goals + away_goals else 0

In [36]:
print("total_shots: "+str(total_shots/total_tested))
print("total_shots_over: "+str(total_shots_over/total_tested))
print("total_shots_under: "+str(total_shots_under/total_tested))
print("\n")
print("total_shots_on_target: "+str(total_shots_on_target/total_tested))
print("total_shots_on_target_over: "+str(total_shots_on_target_over/total_tested))
print("total_shots_on_target_under: "+str(total_shots_on_target_under/total_tested))
print("\n")
print("total_booking_points: "+str(total_booking_points/total_tested))
print("total_booking_points_over: "+str(total_booking_points_over/total_tested))
print("total_booking_points_under: "+str(total_booking_points_under/total_tested))
print("\n")
print("correct_score: "+str(correct_score/total_tested))
print("\n")
print("winner: "+str(winner/total_tested))
print("\n")
print("total_corners: "+str(total_corners/total_tested))
print("total_corners_over: "+str(total_corners_over/total_tested))
print("total_corners_under: "+str(total_corners_under/total_tested))
print("\n")
print("total_fouls: "+str(total_fouls/total_tested))
print("total_fouls_over: "+str(total_fouls_over/total_tested))
print("total_fouls_under: "+str(total_fouls_under/total_tested))
print("\n")
print("goals_over: "+str(goals_over/total_tested))
print("goals_under: "+str(goals_under/total_tested))
print("\n")
print("all_over: "+str(all_over/total_tested))
print("all_under: "+str(all_under/total_tested))

total_shots: 0.07253886010362694
total_shots_over: 0.41968911917098445
total_shots_under: 0.5077720207253886


total_shots_on_target: 0.11917098445595854
total_shots_on_target_over: 0.3549222797927461
total_shots_on_target_under: 0.5259067357512953


total_booking_points: 0.17875647668393782
total_booking_points_over: 0.14766839378238342
total_booking_points_under: 0.6735751295336787


correct_score: 0.11917098445595854


winner: 0.45595854922279794


total_corners: 0.10880829015544041
total_corners_over: 0.3238341968911917
total_corners_under: 0.5673575129533679


total_fouls: 0.07772020725388601
total_fouls_over: 0.3860103626943005
total_fouls_under: 0.5362694300518135


goals_over: 0.22020725388601037
goals_under: 0.5284974093264249


all_over: 0.0051813471502590676
all_under: 0.10362694300518134


### Tuning

Create a function to build and compile a model, with the arguments in it being the values to be update during tuning

In [37]:
def get_mlp_model(hidden_layer_one=13, dropout=0.2, learn_rate=0.01, n_h_layers=1):

	model = tf.keras.models.Sequential()

	# input
	model.add(tf.keras.layers.Dense(15, activation="relu", input_dim=15))

	for i in range(n_h_layers):
		model.add(tf.keras.layers.Dense(hidden_layer_one, activation="relu"))

	# dropout layer to remove redundant nodes
	model.add(tf.keras.layers.Dropout(dropout))
	
	# output
	model.add(tf.keras.layers.Dense(14, activation="relu"))

	model.compile(
		optimizer=tf.keras.optimizers.legacy.Adam(learning_rate=learn_rate),
		loss="mse",
		metrics=["accuracy"])

	return model

Build a model using the *get_mlp_model* function, letting tensorflow use this function when building the models

In [38]:
model = KerasRegressor(model=get_mlp_model, verbose=0, hidden_layer_one=10, learn_rate=0.01, dropout=0.05, n_h_layers=1);

Define the hyperparameter search space, and then generate the search object to be used later, by RandomizedSearchCV

In [39]:
# define a grid of the hyperparameter search space
hidden_layer_one = [10, 15, 30, 50, 100]
learn_rate = [1e-2, 1e-3, 1e-4]
dropout = [0.05, 0.1, 0.2, 0.3, 0.4]
batch_size = [4, 8, 16, 32, 64]
epochs = [10, 50, 100, 150, 200, 500]
n_h_layers = [1, 2, 3, 5]

# create a dictionary from the hyperparameter grid
grid = dict(
	hidden_layer_one=hidden_layer_one,
	learn_rate=learn_rate,
	dropout=dropout,
	batch_size=batch_size,
	epochs=epochs,
	n_h_layers=n_h_layers
)

Build a scoring function for the searcher, average the value of all the *under* stats and return this

In [40]:
highest_average_under_rate = 0

In [41]:
def scoring(estimator, test_x: np.ndarray, test_y: pd.DataFrame) -> float:
	test_y = test_y.to_numpy()
	
	estimator.fit(test_x, test_y);
	y_hat = estimator.predict(test_x)
	
	average_under_rate = 0

	total_shots_under = 0
	total_shots_on_target_under = 0
	total_booking_points_under = 0
	total_corners_under = 0
	total_fouls_under = 0
	goals_under = 0
	all_under = 0

	total_tested = len(y_hat)
	
	for idx, y in enumerate(y_hat):

		home_goals_hat, away_goals_hat, home_shots_hat, away_shots_hat, home_shots_on_target_hat, away_shots_on_target_hat, home_corners_hat, away_corners_hat, home_fouls_hat, away_fouls_hat, home_yellow_cards_hat, away_yellow_cards_hat, home_red_cards_hat, away_red_cards_hat = y
		home_goals, away_goals, home_shots, away_shots, home_shots_on_target, away_shots_on_target, home_corners, away_corners, home_fouls, away_fouls, home_yellow_cards, away_yellow_cards, home_red_cards, away_red_cards = test_y[idx]

		total_shots_under += 1 if np.floor(home_shots_hat)+np.floor(away_shots_hat) < home_shots+away_shots else 0
		total_shots_on_target_under += 1 if np.floor(home_shots_on_target_hat)+np.floor(away_shots_on_target_hat) < home_shots_on_target+away_shots_on_target else 0
		total_booking_points_under += 1 if (np.floor(home_yellow_cards_hat)+np.floor(away_yellow_cards_hat))*10+(np.floor(home_red_cards_hat)+np.floor(away_red_cards_hat))*25 < (home_yellow_cards+away_yellow_cards)*10+(home_red_cards+away_red_cards)*25 else 0
		total_corners_under += 1 if np.floor(home_corners_hat) + np.floor(away_corners_hat) < home_corners + away_corners else 0
		total_fouls_under += 1 if np.floor(home_fouls_hat) + np.floor(away_fouls_hat) < home_fouls+ away_fouls else 0
		goals_under += 1 if np.floor(home_goals_hat) + np.floor(away_goals_hat) < home_goals + away_goals else 0

		average_under_rate += ((total_shots_under/total_tested)+(total_shots_on_target_under/total_tested)+(total_booking_points_under/total_tested)+(total_fouls_under/total_tested)+(total_corners_under/total_tested)+(goals_under/total_tested))/6

	average_under_rate = average_under_rate/total_tested
	
	return average_under_rate

Generate the searcher object, used to iterate over the search parameters and determine the best hyper-parameter setup from the space we've created above

In [42]:
searcher = GridSearchCV(estimator=model, n_jobs=-1, 
	param_grid=grid, scoring=scoring, verbose=4, cv=3)

searchResults = searcher.fit(X_train, y_train)

bestScore = searchResults.best_score_
bestParams = searchResults.best_params_
print("[INFO] best score is {:.2f} using {}".format(bestScore,bestParams))

Fitting 3 folds for each of 9000 candidates, totalling 27000 fits


2024-01-02 10:20:19.038180: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2024-01-02 10:20:19.042354: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2024-01-02 10:20:19.081402: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2024-01-02 10:20:19.114301: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2024-01-02 10:20:19.162270: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2024-01-02 10:20:19.196799: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2024-01-02 10:20:19.246964: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2024-01-02 10:20:19.275070: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


[CV 3/3] END batch_size=4, dropout=0.05, epochs=10, hidden_layer_one=10, learn_rate=0.01, n_h_layers=1;, score=0.292 total time=   5.9s
[CV 2/3] END batch_size=4, dropout=0.05, epochs=10, hidden_layer_one=10, learn_rate=0.01, n_h_layers=2;, score=0.331 total time=   6.0s
[CV 1/3] END batch_size=4, dropout=0.05, epochs=10, hidden_layer_one=10, learn_rate=0.01, n_h_layers=1;, score=0.298 total time=   6.2s
[CV 2/3] END batch_size=4, dropout=0.05, epochs=10, hidden_layer_one=10, learn_rate=0.01, n_h_layers=1;, score=0.312 total time=   6.2s
[CV 3/3] END batch_size=4, dropout=0.05, epochs=10, hidden_layer_one=10, learn_rate=0.01, n_h_layers=2;, score=0.269 total time=   6.3s
[CV 1/3] END batch_size=4, dropout=0.05, epochs=10, hidden_layer_one=10, learn_rate=0.01, n_h_layers=3;, score=0.253 total time=   6.5s
[CV 2/3] END batch_size=4, dropout=0.05, epochs=10, hidden_layer_one=10, learn_rate=0.01, n_h_layers=3;, score=0.260 total time=   6.8s
[CV 1/3] END batch_size=4, dropout=0.05, epochs=

KeyboardInterrupt: 

## Next v  steps

- Tune parameters and layers on the 15 PC model to increase the accuracy
- Create a method to define the accuracy on the test set; strict equality for each of the output values is not necessary as under/over options on the betting platform mean a close approximation of the outcome is all that is required. Success/accuracy measure based on the bet markets available for each game:
	- Total shots
	- Total booking points
	- Correct score
	- Outright
	- Total corners
	- Under/over goals
- Save the model
- Complete the notebook with all annotations and text

RandomSearchCV  
best score is 0.39 using {'n_h_layers': 5, 'learn_rate': 0.0001, 'hidden_layer_one': 15, 'epochs': 100, 'dropout': 0.2, 'batch_size': 64}  
best score is 0.40 using {'n_h_layers': 7, 'learn_rate': 0.0001, 'hidden_layer_one': 15, 'epochs': 1000, 'dropout': 0.75, 'batch_size': 16}  
best score is 0.49 using {'n_h_layers': 3, 'learn_rate': 0.0001, 'hidden_layer_one': 30, 'epochs': 10, 'dropout': 0.3, 'batch_size': 64} 

GridSearchCV  


## Questions

- Scaling y reduces the accuracy of the model by a factor of 6, why?